pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -426,3 +438,242 @@ func (vs *VSphere) Routes() (cloudprovider.Routes, bool) {
 }
+
+// Attaches given virtual disk volume to the compute running kubelet.
+func (vs *VSphere) AttachDisk(vmDiskPath string, nodeName string) (diskID string, diskUUID string, err error) {
+	// Create context
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Create vSphere client
+	c, err := vsphereLogin(vs.cfg, ctx)
+	if err != nil {
+		return "", "", err
+	}
+	defer c.Logout(ctx)
+
+	// Create a new finder
+	f := find.NewFinder(c.Client, true)
+
+	// Fetch and set data center
+	dc, err := f.Datacenter(ctx, vs.cfg.Global.Datacenter)
+	if err != nil {
+		return "", "", err
+	}
+	f.SetDatacenter(dc)
+
+	// Find datastores
+	ds, err := f.Datastore(ctx, vs.cfg.Global.Datastore)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Find virtual machine to attach disk to
+	var vSphereInstance string
+	if nodeName == "" {
+		vSphereInstance = vs.localInstanceID
+	} else {
+		vSphereInstance = nodeName
+	}
+	vm, err := f.VirtualMachine(ctx, vSphereInstance)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Attach disk to the running VM
+	vmDevices, err := vm.Device(ctx)
+	if err != nil {
+		return "", "", err
+	}
+
+	diskController, err := vmDevices.FindDiskController(vs.cfg.Disk.DiskController)
+	if err != nil {
+		if err.Error() == "no available SCSI controller" {
+			newSCSIController, err := vmDevices.CreateSCSIController(vs.cfg.Disk.SCSIControllerType)
+			if err != nil {
+				return "", "", err
+			}
+
+			configNewSCSIController := newSCSIController.(types.BaseVirtualSCSIController).GetVirtualSCSIController()
+			hotAndRemove := true
+			configNewSCSIController.HotAddRemove = &hotAndRemove
+			configNewSCSIController.SharedBus = types.VirtualSCSISharing(types.VirtualSCSISharingNoSharing)
+			err = vm.AddDevice(context.TODO(), newSCSIController)
+			if err != nil {
+				return "", "", err
+			}
+			vmDevices, err = vm.Device(ctx)
+			if err != nil {
+				return "", "", err
+			}
+			diskController, err = vmDevices.FindDiskController(vs.cfg.Disk.DiskController)
+			if err != nil {
+				return "", "", err
+			}
+		} else {
+			return "", "", err
+		}
+	}
+
+	disk := vmDevices.CreateDisk(diskController, ds.Reference(), vmDiskPath)
+	backing := disk.Backing.(*types.VirtualDiskFlatVer2BackingInfo)
+	backing.DiskMode = string(types.VirtualDiskModeIndependent_persistent)
+	disk = vmDevices.ChildDisk(disk)
+
+	err = vm.AddDevice(ctx, disk)
+	if err != nil {
+		return "", "", err
+	}
+
+	vmDevices, err = vm.Device(ctx)
+	if err != nil {
+		return "", "", err
+	}
+
+	devices := vmDevices.SelectByType(disk)
+	newDevice := devices[len(devices)-1]
+
+	diskUUID = getVirtualDiskUUID(newDevice)
+
+	return devices.Name(newDevice), diskUUID, nil
+}
+
+func getVirtualDiskUUID(newDevice types.BaseVirtualDevice) string {
+	vd := newDevice.GetVirtualDevice()
+
+	if b, ok := vd.Backing.(*types.VirtualDiskFlatVer2BackingInfo); ok {
+		uuidWithNoHypens := strings.Replace(b.Uuid, "-", "", -1)
+		return uuidWithNoHypens
+	}
+	return ""
```

Comment:
if disk UUID fails, we should return error with this function and caller (attach disk function) should fail and log error.  Without this information, mount will also fail and would be harder to diagnose if we don't log/fail here.